tools/calyx-ffi-macro/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+    let comp_name =
+        syn::parse_str::<syn::LitStr>(&format!("\"{}\"", comp.name))
+            .expect("failed to turn quoted name into string");
+    let comp_path = syn::parse_str::<syn::LitStr>(&format!(
+        "\"{}\"",
+        path.to_string_lossy()
+    ))
+    .expect("failed to turn quoted path into string");
+
+    let backend_macro = args.backend;
+    let mut input_names = Vec::new();
+    let mut output_names = Vec::new();
+    let mut fields = vec![];
+    let mut default_field_inits = vec![];
+    let mut getters = vec![];
+    let mut setters = vec![];
+    let mut width_getters = vec![];
+    let mut port_names = vec![];
+
+    for port in comp.signature.borrow().ports() {
+        let port_name_str = port.borrow().name.to_string();
+        let port_name = syn::parse_str::<syn::Ident>(&port_name_str)
+            .expect("failed to turn port name into identifier");
+
+        port_names.push(port_name.clone());
+
+        let port_width = port.borrow().width;
+        let width_getter = format_ident!("{}_width", port_name);
+        width_getters.push(quote! {
+            pub const fn #width_getter() -> u64 {
+                #port_width
+            }
+        });
+
+        default_field_inits.push(quote! {
+            #port_name: calyx_ffi::value_from_u64::<#port_width>(0)
+        });
+
+        // idk why input output ports are being flipped??
+        match port.borrow().direction.reverse() {
+            calyx_ir::Direction::Input => {
+                let setter = format_ident!("set_{}", port_name);
+                fields.push(quote! {
+                    pub #port_name: calyx_ffi::Value<#port_width>
+                });
+                setters.push(quote! {
+                    pub fn #setter(&mut self, value: u64) {
+                        self.#port_name = calyx_ffi::value_from_u64::<#port_width>(value);
+                    }
+                });
+                input_names.push(port_name);
+            }
+            calyx_ir::Direction::Output => {
+                fields.push(quote! {
+                    #port_name: calyx_ffi::Value<#port_width>
+
+                });
+
+                let bitvec_getter = format_ident!("{}_bits", port_name);
+
+                getters.push(quote! {
+                    pub fn #port_name(&self) -> u64 {
+                        interp::BitVecOps::to_u64(&self.#port_name).expect("port value wider than 64 bits")
+                    }
+
+                    pub const fn #bitvec_getter(&self) -> &calyx_ffi::Value<#port_width> {
+                        &self.#port_name
+                    }
+                });
+                output_names.push(port_name);
+            }
+            calyx_ir::Direction::Inout => {
+                todo!("inout ports not supported yet")
+            }
+        }
+    }
+
+    let struct_def = quote! {
+        struct #name {
+            #(#fields,)*
+            user_data: std::mem::MaybeUninit<#backend_macro!(@user_data_type)>
```

Comment:
What is this for and why is it `MaybeUninit`?

Reply:
It stores whatever data the backend needs it to store. It's `MaybeUninit` because one `CalyxFFI` context can create multiple components, so it has to be dynamically independent of the backend.

Reply:
I'll put a comment to explain its purpose.